examples_tests/common/CommonAPI.h

```diff
@@ -443,128 +681,256 @@ class CommonAPI
+	static nbl::core::smart_refctd_ptr<nbl::video::ISwapchain> createSwapchain(
+		nbl::video::E_API_TYPE api_type,
+		uint32_t width,
 		uint32_t height,
 		uint32_t imageCount,
 		const nbl::core::smart_refctd_ptr<nbl::video::ILogicalDevice>& device,
 		const nbl::core::smart_refctd_ptr<nbl::video::ISurface>& surface,
-		nbl::video::ISurface::E_PRESENT_MODE presentMode)
+		nbl::video::ISurface::E_PRESENT_MODE requestedPresentMode,
+		nbl::video::ISurface::SFormat requestedSurfaceFormat,
+		const GPUInfo& gpuInfo)
 	{
 		using namespace nbl;
-		video::ISwapchain::SCreationParams sc_params;
+
+		nbl::asset::E_SHARING_MODE imageSharingMode;
+		if (gpuInfo.queueFamilyProps.graphics.index == gpuInfo.queueFamilyProps.present.index)
+			imageSharingMode = asset::ESM_EXCLUSIVE;
+		else
+			imageSharingMode = asset::ESM_CONCURRENT;
+		
+		nbl::video::ISurface::SFormat surfaceFormat;
+
+		if(api_type == video::EAT_VULKAN)
+		{
+			uint32_t found_format_and_colorspace = ~0u;
+			uint32_t found_format = ~0u;
+			for(uint32_t i = 0; i < gpuInfo.availableSurfaceFormats.size(); ++i)
+			{
+				const auto & supportedFormat = gpuInfo.availableSurfaceFormats[i];
+				if(requestedSurfaceFormat.format == supportedFormat.format)
+				{
+					if(found_format == ~0u)
+						found_format = i;
+					if(requestedSurfaceFormat.colorSpace.eotf == supportedFormat.colorSpace.eotf && requestedSurfaceFormat.colorSpace.primary == supportedFormat.colorSpace.primary)
+					{
+						found_format_and_colorspace = i;
+						break;
+					}
+				}
+			}
+		
+			if(found_format_and_colorspace != ~0u)
+			{
+				surfaceFormat = gpuInfo.availableSurfaceFormats[found_format_and_colorspace];
+			}
+			else if(found_format != ~0u) // fallback
+			{
+				assert(false && "Fallback: requested 'colorspace' is not supported.");
+				surfaceFormat = gpuInfo.availableSurfaceFormats[found_format];
+			}
+			else
+			{
+				assert(false && "Fallback: requested 'format' and 'colorspace' is not supported.");
+				surfaceFormat = gpuInfo.availableSurfaceFormats[0];
+			}
+
+			bool presentModeSupported = (gpuInfo.availablePresentModes & requestedPresentMode) != 0;
+			assert(presentModeSupported);
+			if(!presentModeSupported) // fallback 
+			{
+				requestedPresentMode = nbl::video::ISurface::E_PRESENT_MODE::EPM_FIFO;
+				assert(false && "Fallback: requested 'present mode' is not supported");
+			}
+		}
+		else
+		{
+			surfaceFormat = requestedSurfaceFormat;
+		}
+
+		video::ISwapchain::SCreationParams sc_params = {};
 		sc_params.width = width;
 		sc_params.height = height;
 		sc_params.arrayLayers = 1u;
 		sc_params.minImageCount = imageCount;
-		sc_params.presentMode = presentMode;
+		sc_params.presentMode = requestedPresentMode;
+		sc_params.imageUsage = static_cast<asset::IImage::E_USAGE_FLAGS>(asset::IImage::EUF_COLOR_ATTACHMENT_BIT | asset::IImage::EUF_STORAGE_BIT | asset::IImage::EUF_TRANSFER_DST_BIT);;
```

Comment:
why transfer DST?

For screenshots it should be SRC

Reply:
Probably @Erfan-Ahmadi needs to see this

Reply:
Because of `vkCmdBlitImage`
and EUF_STORAGE_BIT doesn't seem to be well supported so I removed it from my branch, I will also add `TRANSFER_SRC`

Reply:
I think Vulkan's mindset is more like this: if you want to render to the swapchain, you must either have that as COLOR_ATTACHMENT or write to another image (in compute or whatever) with supported usage flags and VkCmdBlit to the swapchain image.
Example 42. runs fine without it.